File: scripts/crypto.ts
import {
  createCipheriv,
  createDecipheriv,
  pbkdf2Sync,
  randomBytes,
} from "node:crypto";

const KDF_ITERATIONS = 250_000;
const KEY_LENGTH = 32;
const SALT_LENGTH = 16;
const IV_LENGTH = 12;
const AUTH_TAG_LENGTH = 16;

export type CryptoFormat = {
  magic: string;
};

export function encrypt(
  plaintext: Buffer,
  passphrase: string,
  format: CryptoFormat,
): Buffer {
  const salt = randomBytes(SALT_LENGTH);
  const iv = randomBytes(IV_LENGTH);
  const key = pbkdf2Sync(
    passphrase,
    salt,
    KDF_ITERATIONS,
    KEY_LENGTH,
    "sha256",
  );

  const cipher = createCipheriv("aes-256-gcm", key, iv);
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const authTag = cipher.getAuthTag();

  return Buffer.concat([
    Buffer.from(format.magic),
    salt,
    iv,
    authTag,
    ciphertext,
  ]);
}

export function decrypt(
  payload: Buffer,
  passphrase: string,
  format: CryptoFormat,
): Buffer {
  const minLength =
    format.magic.length + SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH;

  if (payload.length < minLength) {
    throw new Error("Encrypted payload is malformed or truncated.");
  }

  const magic = payload.subarray(0, format.magic.length).toString("utf8");
  if (magic !== format.magic) {
    throw new Error(
      `Unexpected format magic: ${magic} (expected ${format.magic})`,
    );
  }

  let offset = format.magic.length;
  const salt = payload.subarray(offset, offset + SALT_LENGTH);
  offset += SALT_LENGTH;
  const iv = payload.subarray(offset, offset + IV_LENGTH);
  offset += IV_LENGTH;
  const authTag = payload.subarray(offset, offset + AUTH_TAG_LENGTH);
  offset += AUTH_TAG_LENGTH;
  const ciphertext = payload.subarray(offset);

  const key = pbkdf2Sync(
    passphrase,
    salt,
    KDF_ITERATIONS,
    KEY_LENGTH,
    "sha256",
  );

  const decipher = createDecipheriv("aes-256-gcm", key, iv);
  decipher.setAuthTag(authTag);

  try {
    return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  } catch {
    throw new Error(
      "Failed to decrypt and authenticate payload. Check passphrase and integrity.",
    );
  }
}

File: scripts/crypto.py
"""AES-256-GCM encryption/decryption with PBKDF2 key derivation.

Format-compatible with the original TypeScript implementation.
Requires the `cryptography` package: uv add cryptography
"""

from __future__ import annotations

import hashlib
import os
from dataclasses import dataclass

KDF_ITERATIONS = 250_000
KEY_LENGTH = 32
SALT_LENGTH = 16
IV_LENGTH = 12
AUTH_TAG_LENGTH = 16


@dataclass
class CryptoFormat:
    magic: str


def _derive_key(passphrase: str, salt: bytes) -> bytes:
    return hashlib.pbkdf2_hmac(
        "sha256", passphrase.encode(), salt, KDF_ITERATIONS, dklen=KEY_LENGTH
    )


def encrypt(plaintext: bytes, passphrase: str, fmt: CryptoFormat) -> bytes:
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    except ImportError as exc:
        raise RuntimeError(
            "The 'cryptography' package is required for encryption. Install with: uv add cryptography"
        ) from exc

    salt = os.urandom(SALT_LENGTH)
    iv = os.urandom(IV_LENGTH)
    key = _derive_key(passphrase, salt)

    aesgcm = AESGCM(key)
    # AESGCM.encrypt returns ciphertext + auth_tag (16 bytes appended)
    ct_with_tag = aesgcm.encrypt(iv, plaintext, None)
    ciphertext = ct_with_tag[:-AUTH_TAG_LENGTH]
    auth_tag = ct_with_tag[-AUTH_TAG_LENGTH:]

    return fmt.magic.encode() + salt + iv + auth_tag + ciphertext


def decrypt(payload: bytes, passphrase: str, fmt: CryptoFormat) -> bytes:
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    except ImportError as exc:
        raise RuntimeError(
            "The 'cryptography' package is required for decryption. Install with: uv add cryptography"
        ) from exc

    magic_len = len(fmt.magic)
    min_length = magic_len + SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH

    if len(payload) < min_length:
        raise RuntimeError("Encrypted payload is malformed or truncated.")

    magic = payload[:magic_len].decode("utf-8")
    if magic != fmt.magic:
        raise RuntimeError(f"Unexpected format magic: {magic} (expected {fmt.magic})")

    offset = magic_len
    salt = payload[offset : offset + SALT_LENGTH]
    offset += SALT_LENGTH
    iv = payload[offset : offset + IV_LENGTH]
    offset += IV_LENGTH
    auth_tag = payload[offset : offset + AUTH_TAG_LENGTH]
    offset += AUTH_TAG_LENGTH
    ciphertext = payload[offset:]

    key = _derive_key(passphrase, salt)
    aesgcm = AESGCM(key)

    # Reconstruct ciphertext + tag as AESGCM.decrypt expects
    ct_with_tag = ciphertext + auth_tag

    try:
        return aesgcm.decrypt(iv, ct_with_tag, None)
    except Exception as exc:
        raise RuntimeError(
            "Failed to decrypt and authenticate payload. Check passphrase and integrity."
        ) from exc
